File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/SearchRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.data.model.SearchResponse

/** Hybrid search (`GET /search?q=...`). */
open class SearchRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    open suspend fun search(query: String): SearchResponse {
        val trimmed = query.trim()
        val path = Endpoints.withQuery(Endpoints.SEARCH, "q" to trimmed)
        return apiClient.get(path, token = tokenStore.requireAccessToken())
    }
}

File: bubbler_app/lib/data/repositories/search_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/search.dart';

/// Hybrid search (`GET /search?q=...`).
class SearchRepository {
  SearchRepository(this._client);

  final ApiClient _client;

  /// Keyword/topic/username hits plus semantic related posts.
  Future<SearchResponse> search(String query) async {
    final trimmed = query.trim();
    final data = await _client.authorizedRequest(
      path: Endpoints.search,
      queryParameters: {'q': trimmed},
    );
    return SearchResponse.fromJson(_asJsonMap(data));
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}
